### PR TITLE
Support launch query engine asynchronously during opening session

### PR DESCRIPTION
**Pull request by turboFei**
### _Why are the changes needed?_

Support launch query engine asynchronously during opening session, make it more user-friendly.

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by turboFei** [comment]
https://issues.apache.org/jira/browse/THRIFT-563
will use TMultiplexedProcessor

File: kyuubi-service-rpc/pom.xml [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <groupId>org.apache.kyuubi</groupId>
+        <artifactId>kyuubi-parent</artifactId>
+        <version>1.4.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>kyuubi-service-rpc</artifactId>
```
Comment (ulysses-you) [inline kyuubi-service-rpc/pom.xml:29]: does it mean we don't need `hive-service` dependency any more?

Comment (turboFei) [reply kyuubi-service-rpc/pom.xml:29]: we still need it, kyuubi thrift rpc is a supplement.

File: kyuubi-service-rpc/pom.xml [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+                        <artifactId>maven-antrun-plugin</artifactId>
+                        <executions>
+                            <execution>
+                                <id>generate-thrift-sources</id>
+                                <phase>generate-sources</phase>
+                                <configuration>
+                                    <target>
+                                        <taskdef name="for" classname="net.sf.antcontrib.logic.ForTask"
+                                                 classpathref="maven.plugin.classpath" />
+                                        <property name="thrift.args" value="-I ${thrift.home} --gen java:beans,generated_annotations=undated --gen cpp --gen php --gen py --gen rb"/>
```
Comment (pan3793) [inline kyuubi-service-rpc/pom.xml:121]: Do we need `--gen cpp --gen php --gen py --gen rb`?

File: pom.xml [hunk 3]
```diff
@@ -146,6 +147,7 @@
         <jars.target.dir>${project.build.directory}/scala-${scala.binary.version}/jars</jars.target.dir>
 
         <maven.plugin.antrun.version>3.0.0</maven.plugin.antrun.version>
+        <ant.contrib.version>1.0b3</ant.contrib.version>
```
Comment (pan3793) [inline pom.xml:150]: `maven.plugin.ant.contrib.version`

File: kyuubi-service-rpc/pom.xml [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+    <artifactId>kyuubi-service-rpc</artifactId>
+    <name>Kyuubi Project Service RPC</name>
+    <packaging>jar</packaging>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.thrift</groupId>
+            <artifactId>libthrift</artifactId>
+            <version>${hive.jdbc.thrift.version}</version>
```
Comment (pan3793) [inline kyuubi-service-rpc/pom.xml:37]: We should NOT use `hive.jdbc...` here, it's dedicated for `kyuubi-hive-jdbc` module

File: kyuubi-service-rpc/pom.xml (hunk 2 above)
Comment (turboFei) [reply kyuubi-service-rpc/pom.xml:121]: I think we can reserve them, there is no maintain efforts.

**Comment by turboFei** [comment]
TODO:

1. add contribution guide for rebuild thrift classes
2. Extend the KyuubiConnect to use kyuubi thrift API

File: kyuubi-service-rpc/pom.xml [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+                                    </rules>
+                                    <fail>true</fail>
+                                </configuration>
+                            </execution>
+                        </executions>
+                    </plugin>
+                    <plugin>
+                        <groupId>com.google.code.maven-replacer-plugin</groupId>
+                        <artifactId>replacer</artifactId>
+                        <version>1.5.3</version>
```
Comment (pan3793) [inline kyuubi-service-rpc/pom.xml:167]: extract to maven property

File: kyuubi-service-rpc/pom.xml [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+    <artifactId>kyuubi-service-rpc</artifactId>
+    <name>Kyuubi Project Service RPC</name>
+    <packaging>jar</packaging>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.thrift</groupId>
+            <artifactId>libthrift</artifactId>
+            <version>${kyuubi.service.rpc.thrift.version}</version>
```
Comment (pan3793) [inline kyuubi-service-rpc/pom.xml:37]: any reason create new one rather than using `${thrift.version}`?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/ThriftBinaryFrontendService.scala [hunk 7]
```diff
@@ -42,14 +45,14 @@ abstract class ThriftBinaryFrontendService(name: String)
   import KyuubiConf._
   import ThriftBinaryFrontendService._
 
-  private var server: Option[TServer] = None
-  private var serverThread: Thread = _
+  protected var server: Option[TServer] = None
+  protected var serverThread: Thread = _
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/service/ThriftBinaryFrontendService.scala:49]: nit

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/InitEngine.scala [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  private def openEngineSession(): Unit = {
+    setState(OperationState.RUNNING)
+    try {
+      session.openEngineSession()
+      client = session.sessionManager.operationManager.getThriftClient(session.handle)
+      setState(OperationState.FINISHED)
+    } catch {
+      case e: Throwable =>
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/operation/InitEngine.scala:49]: we can onError here？

Comment (turboFei) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/operation/InitEngine.scala]: thanks

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 9]
```diff
@@ -657,6 +657,20 @@ object KyuubiConf {
...
+      .doc("When opening kyuubi session, whether to init backend engine synchronously")
+      .version("1.4.0")
+      .booleanConf.
+      createWithDefault(true)
+
+  val SESSION_ENGINE_STARTUP_CHECK_INTERVAL: ConfigEntry[Long] =
+    buildConf("session.engine.startup.check.interval")
+      .doc("The interval to check whether backend engine is ready during engine startup phase.")
+      .version("1.4.0")
+      .longConf
```
Comment (ulysses-you) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:671]: timeConf ?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 10]
```diff
@@ -64,44 +69,76 @@ class KyuubiSessionImpl(
...
+      sessionEvent.totalOperations += 1
+    }
     super.runOperation(operation)
   }
 
+  private def waitForEngineInitOpFinished(): Unit = {
+    if (!engineInitFinished) {
+      Option(engineInitOp).foreach { op =>
+        while (!OperationState.isTerminal(op.getStatus.state)) {
+          info("The engine init operation has not finished")
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:126]: I think debug is enough and we can add a new info logging before and after `while`.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 11]
```diff
@@ -657,6 +657,20 @@ object KyuubiConf {
...
       .createWithDefault(10)
 
+  val SESSION_ENGINE_SYNC_INIT: ConfigEntry[Boolean] =
+    buildConf("session.engine.sync.init")
+      .doc("When opening kyuubi session, whether to init backend engine synchronously")
+      .version("1.4.0")
+      .booleanConf.
+      createWithDefault(true)
+
+  val SESSION_ENGINE_STARTUP_CHECK_INTERVAL: ConfigEntry[Long] =
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:667]: The current engine startup check is hard coded, maybe we can just follow that just for now

File: kyuubi-common/src/main/scala/org/apache/kyuubi/operation/OperationType.scala [hunk 12]
```diff
@@ -57,6 +58,7 @@ object OperationType extends Enumeration {
       case GET_TABLE_TYPES => TOperationType.GET_TABLE_TYPES
       case GET_COLUMNS => TOperationType.GET_COLUMNS
       case GET_FUNCTIONS => TOperationType.GET_FUNCTIONS
+      case INIT_ENGINE => TOperationType.UNKNOWN
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/operation/OperationType.scala:61]: How about LAUNCH_ENGINE？

We can add some comments here?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala (hunk 10 above)
Comment (ulysses-you) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:126]: and we can aslo record the elapsed time

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 13]
```diff
@@ -64,44 +69,76 @@ class KyuubiSessionImpl(
...
   override def open(): Unit = {
     MetricsSystem.tracing { ms =>
       ms.incCount(CONN_TOTAL)
       ms.incCount(MetricRegistry.name(CONN_OPEN, user))
     }
-    withZkClient(sessionConf) { zkClient =>
-      val (host, port) = engine.getOrCreate(zkClient)
-      openSession(host, port)
-    }
+    _handle = SessionHandle(protocol)
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:81]: this seems a breaking change? @ulysses-you ?

Comment (ulysses-you) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:81]: yeah, we should keep it as null until we opened the enign session. so we can use engien session handle as kyuubi server side session handle.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/InitEngine.scala [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  private def openEngineSession(): Unit = {
+    setState(OperationState.RUNNING)
+    try {
+      session.openEngineSession()
+      client = session.sessionManager.operationManager.getThriftClient(session.handle)
+      setState(OperationState.FINISHED)
+    } catch onError()
+  }
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/operation/InitEngine.scala:49]: we can move it inside runInternal

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala (hunk 13 above)
Comment (yaooqinn) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:81]: > yeah, we should keep it as null until we opened the enign session. so we can use engien session handle as kyuubi server side session handle.

this will not work as in this feature, the session handle shall be returned ASAP

Comment (yaooqinn) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:81]: in this case we can turn it back to a `val` to keep the old way

Comment (ulysses-you) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:81]: if so , we may need reocrd two session handle in event log

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/InitEngine.scala [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+
+  override protected def runInternal(): Unit = {
+    val asyncOperation: Runnable = () => openEngineSession()
+    try {
+      val opHandle = session.sessionManager.submitBackgroundOperation(asyncOperation)
+      setBackgroundHandle(opHandle)
+    } catch onError("submitting open engine operation in background, request rejected")
+
+    if (!shouldRunAsync) getBackgroundHandle.get()
```
[the remark this answers is not in the document]
Comment (turboFei) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/operation/InitEngine.scala:58]: sorry, what's the meaning?

[the remark this answers is not in the document]
Comment (yaooqinn) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/operation/InitEngine.scala:58]: mistake

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala [hunk 16]
```diff
@@ -26,18 +26,18 @@ import org.apache.thrift.TException
...
 
-abstract class KyuubiOperation(
-    opType: OperationType,
-    session: Session,
-    client: KyuubiSyncThriftClient) extends AbstractOperation(opType, session) {
+abstract class KyuubiOperation(opType: OperationType, session: Session)
+  extends AbstractOperation(opType, session) {
+
+  protected def client = session.asInstanceOf[KyuubiSessionImpl].sessionManager
+    .operationManager.getThriftClient(session.handle)
```
Comment (turboFei) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala:40]: move the client here.

As it might be null before the engine session is ready.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala [hunk 17]
```diff
@@ -92,7 +92,7 @@ abstract class KyuubiOperation(
       setState(OperationState.CANCELED)
       if (_remoteOpHandle != null) {
         try {
-          client.cancelOperation(_remoteOpHandle)
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala:95]: Can `_remoteOpHandle !=null` here mean that client is not null too?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala [hunk 18]
```diff
@@ -112,7 +112,7 @@ abstract class KyuubiOperation(
         }
 
         try {
-          client.closeOperation(_remoteOpHandle)
+          Option(client).foreach(_.closeOperation(_remoteOpHandle))
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala:115]: ditto

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala [hunk 19]
```diff
@@ -26,18 +26,18 @@ import org.apache.thrift.TException
...
 import org.apache.kyuubi.util.ThriftUtils
 
-abstract class KyuubiOperation(
-    opType: OperationType,
-    session: Session,
-    client: KyuubiSyncThriftClient) extends AbstractOperation(opType, session) {
+abstract class KyuubiOperation(opType: OperationType, session: Session)
+  extends AbstractOperation(opType, session) {
+
+  protected def client = session.asInstanceOf[KyuubiSessionImpl].sessionManager
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala:39]: maybe we call expose `session.client` here ?

**Comment by yaooqinn** [comment]
Looks good to me in general.

Can we add more tests for this feature

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 20]
```diff
@@ -657,6 +657,13 @@ object KyuubiConf {
       .checkValue(_ > 0, "the maximum must be positive integer.")
       .createWithDefault(10)
 
+  val SESSION_ENGINE_SYNC_INIT: ConfigEntry[Boolean] =
+    buildConf("session.engine.sync.init")
+      .doc("When opening kyuubi session, whether to init backend engine synchronously")
+      .version("1.4.0")
+      .booleanConf.
+      createWithDefault(true)
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:665]: nit

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 21]
```diff
@@ -54,62 +54,103 @@ class KyuubiSessionImpl(
...
+      sessionEvent.totalOperations += 1
+    }
     super.runOperation(operation)
   }
 
+  private def waitForEngineInitOpFinished(): Unit = {
+    if (!engineInitFinished) {
+      Option(engineInitOp).foreach { op =>
+        val waitingStartTime = System.currentTimeMillis()
+        info(s"Starting to wait the engine init operation finished")
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:127]: info  -> logSessionInfo

File: kyuubi-server/src/main/scala/org/apache/kyuubi/events/KyuubiSessionEvent.scala [hunk 22]
```diff
@@ -23,7 +23,7 @@ import org.apache.kyuubi.server.KyuubiServer
 import org.apache.kyuubi.session.KyuubiSessionImpl
 
 /**
- * @param sessionId server session id
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/events/KyuubiSessionEvent.scala:26]: we can keep the name and name `engineSessionId` as `remoteSessionId` just like other events

**Review by ulysses-you — APPROVED** [review]

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 23]
```diff
@@ -54,62 +54,103 @@ class KyuubiSessionImpl(
...
-  private def openSession(host: String, port: Int): Unit = {
-    val passwd = Option(password).filter(_.nonEmpty).getOrElse("anonymous")
-    val loginTimeout = sessionConf.get(ENGINE_LOGIN_TIMEOUT).toInt
-    val requestTimeout = sessionConf.get(ENGINE_REQUEST_TIMEOUT).toInt
-    transport = PlainSASLHelper.getPlainTransport(
-      user, passwd, new TSocket(host, port, requestTimeout, loginTimeout))
-    if (!transport.isOpen) {
-      transport.open()
-      logSessionInfo(s"Connected to engine [$host:$port]")
+  private[kyuubi] def openEngineSession(): Unit = synchronized {
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:91]: is synchronized necessary?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 24]
```diff
@@ -54,62 +54,103 @@ class KyuubiSessionImpl(
...
     MetricsSystem.tracing { ms =>
       ms.incCount(CONN_TOTAL)
       ms.incCount(MetricRegistry.name(CONN_OPEN, user))
     }
-    withZkClient(sessionConf) { zkClient =>
-      val (host, port) = engine.getOrCreate(zkClient)
-      openSession(host, port)
-    }
-    // we should call super.open after kyuubi session is already opened
+    _handle = SessionHandle(protocol)
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:81]: make handle and _ handle a single val？

File: kyuubi-server/src/test/scala/org/apache/kyuubi/operation/KyuubiOperationPerConnectionSuite.scala [hunk 25]
```diff
@@ -97,4 +97,39 @@ class KyuubiOperationPerConnectionSuite extends WithKyuubiServer with HiveJDBCTe
         "Caused by: java.net.SocketException: Broken pipe (Write failed)")
     }
   }
+
```
Comment (yaooqinn) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/operation/KyuubiOperationPerConnectionSuite.scala:100]: can we also add a test to test in async mode still work for USER mode to share the same engine？

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 26]
```diff
@@ -54,62 +54,103 @@ class KyuubiSessionImpl(
...
+      transport = PlainSASLHelper.getPlainTransport(
+        user, passwd, new TSocket(host, port, requestTimeout, loginTimeout))
+      if (!transport.isOpen) {
+        transport.open()
+        logSessionInfo(s"Connected to engine [$host:$port]")
+      }
+      _client = new KyuubiSyncThriftClient(new TBinaryProtocol(transport))
+      val engineSessionHandle = _client.openSession(protocol, user, passwd, normalizedConf)
+      logSessionInfo(s"Opened engine session[$engineSessionHandle]")
+      sessionManager.operationManager.setConnection(handle, _client)
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:106]: is this still necessary？

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 27]
```diff
@@ -54,62 +54,103 @@ class KyuubiSessionImpl(
   }
 
   val engine: EngineRef = new EngineRef(sessionConf, user)
+  val engineSyncInit = sessionConf.get(SESSION_ENGINE_SYNC_INIT)
+  var engineInitOp: Operation = _
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:58]: private val engineInitOp = essionManager.operationManager.newLaunchEngineOperation(this)

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 28]
```diff
@@ -54,62 +54,103 @@ class KyuubiSessionImpl(
...
-  private var client: KyuubiSyncThriftClient = _
+  private var _client: KyuubiSyncThriftClient = _
+  def client: KyuubiSyncThriftClient = _client
 
   private var _handle: SessionHandle = _
   override def handle: SessionHandle = _handle
 
+  @volatile
+  private var _initEngineOpHandle: OperationHandle = _
+  def initEngineOpHandle: OperationHandle = _initEngineOpHandle
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:74]: we can remove these two as we have engineInitOp already

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala [hunk 29]
```diff
@@ -42,7 +42,7 @@ class KyuubiOperationManager private (name: String) extends OperationManager(nam
     super.initialize(conf)
   }
 
-  private def getThriftClient(sessionHandle: SessionHandle): KyuubiSyncThriftClient = {
+  private[operation] def getThriftClient(sessionHandle: SessionHandle): KyuubiSyncThriftClient = {
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala:45]: anywhere else used?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/LaunchEngine.scala [hunk 30]
```diff
@@ -0,0 +1,59 @@
...
+
+package org.apache.kyuubi.operation
+
+import org.apache.kyuubi.operation.log.OperationLog
+import org.apache.kyuubi.session.KyuubiSessionImpl
+
+class LaunchEngine(session: KyuubiSessionImpl) extends
+  KyuubiOperation(OperationType.LAUNCH_ENGINE, session) {
+
+  private final val _operationLog: OperationLog = {
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/operation/LaunchEngine.scala:26]: when shouldRunAsync we don't need _operationLog, right?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/LaunchEngine.scala [hunk 31]
```diff
@@ -0,0 +1,59 @@
...
+    }
+    try {
+      val opHandle = session.sessionManager.submitBackgroundOperation(asyncOperation)
+      setBackgroundHandle(opHandle)
+    } catch onError("submitting open engine operation in background, request rejected")
+
+    if (!shouldRunAsync) getBackgroundHandle.get()
+  }
+
+  override def shouldRunAsync: Boolean = !session.engineSyncInit
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/operation/LaunchEngine.scala:58]: can we move this the constructor like ExecuteStatement?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 32]
```diff
@@ -657,6 +657,13 @@ object KyuubiConf {
       .checkValue(_ > 0, "the maximum must be positive integer.")
       .createWithDefault(10)
 
+  val SESSION_ENGINE_SYNC_INIT: ConfigEntry[Boolean] =
+    buildConf("session.engine.sync.init")
+      .doc("When opening kyuubi session, whether to init backend engine synchronously")
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:662]: When false, the Kyuubi server will set up the connection with the client without delay as the backend engine will be created asynchronously if it is absent.

**Review by yaooqinn — APPROVED** [review]

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 33]
```diff
@@ -42,7 +42,7 @@ class KyuubiSessionImpl(
     password: String,
     ipAddress: String,
     conf: Map[String, String],
-    sessionManager: KyuubiSessionManager,
+    override val sessionManager: KyuubiSessionManager,
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:45]: unnecessary change?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 34]
```diff
@@ -657,6 +657,13 @@ object KyuubiConf {
...
+      .version("1.4.0")
+      .booleanConf
+      .createWithDefault(true)
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:665]: Can we also add a `todo` here to make it `false` by default(maybe we shall rename the conf to xx.launch.async and set false now)

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala [hunk 35]
```diff
@@ -135,8 +127,12 @@ class KyuubiOperationManager private (name: String) extends OperationManager(nam
...
       schemaName: String,
       functionName: String): Operation = {
-    val client = getThriftClient(session.handle)
-    val operation = new GetFunctions(session, client, catalogName, schemaName, functionName)
+    val operation = new GetFunctions(session, catalogName, schemaName, functionName)
+    addOperation(operation)
+  }
+
+  def newLaunchEngineOperation(session: KyuubiSessionImpl): Operation = {
+    val operation = new LaunchEngine(session)
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperationManager.scala:135]: when do we close it?

**Comment by turboFei** [comment]
thanks for all your review.
@yaooqinn  @ulysses-you  @pan3793
